core/src/main/scala/kafka/tools/StorageTool.scala

```diff
@@ -109,6 +111,52 @@ object StorageTool extends Logging {
     }
   }
 
+  private def validateMetadataVersion(metadataVersion: MetadataVersion, config: Option[KafkaConfig]): Unit = {
+    if (!metadataVersion.isKRaftSupported) {
+      throw new TerseFailure(s"Must specify a valid KRaft metadata.version of at least ${MetadataVersion.IBP_3_0_IV0}.")
+    }
+    if (!metadataVersion.isProduction) {
+      if (config.get.unstableMetadataVersionsEnabled) {
+        System.out.println(s"WARNING: using pre-production metadata.version $metadataVersion.")
+      } else {
+        throw new TerseFailure(s"The metadata.version $metadataVersion is not ready for production use yet.")
+      }
+    }
+  }
+
+  private[tools] def generateFeatureRecords(metadataRecords: ArrayBuffer[ApiMessageAndVersion],
+                                            metadataVersion: MetadataVersion,
+                                            specifiedFeatures: Map[String, java.lang.Short],
+                                            allFeatures: List[Features],
+                                            usesVersionDefault: Boolean): Unit = {
+    // If we are using --version-default, the default is based on the metadata version.
+    val metadataVersionForDefault = if (usesVersionDefault) Optional.of(metadataVersion) else Optional.empty[MetadataVersion]()
```

Comment:
We do not. This was changed yesterday and the code was as you say in the comment. Here is the behavior as it stands now:

> Do we get the same result by passing in that metadataVersion to feature.defaultValue()

If we follow the protocol of creating a new MV for each new feature and making them production ready at the same time then the answer to your question is yes. If we want to codify this and require a new MV (used only for mapping a default) for every new feature to be created when we mark the feature as production ready, I can switch it back to how it was yesterday.

Reply:
If we want to change it back, I will also update the comments in the metadataVersionMapping method as it will not be correct.

Reply:
Ok sorry to be a little all over the place here. I think we should have semantics where on each new feature version released to production, we have a MV released to production. 

In this case, whether we specify latest production (as it did before) or use an empty optional to specify the latest production feature SHOULD be equivalent. The only case it is not is if someone improperly doesn't set the metadataMapping correctly. There isn't a great way to enforce that (I can do so via a test), but I guess the question is whether we prefer the empty approach that ensures the latest features are provided OR if we prefer the latest production metadata approach that is simpler but may risk not picking up features if folks implement the method incorrectly/don't update the MV.

Reply:
> If we follow the protocol of creating a new MV for each new feature and making them production ready at the same time then the answer to your question is yes.

I thought that's the model being implemented since each FeatureVersion needs to define `bootstrapMetadataVersion()`?

Another thing is that if we follow this model, it would be inconvenient for each feature to maintain its own latestProductionVersion and make sure that it's consistent with the LATEST_PRODUCTION in MV. Perhaps it's simpler to just maintain the latestProductionVersion in MV? Then through `bootstrapMetadataVersion()`, we will know whether a feature level is production ready or not.

Reply:
Ok -- so if I understand correctly, the request is to remove latest production per feature and to simply mark as production ready if the MV that corresponds to it is production ready?

The only case where this is tricky is when we use the --feature flag and we need to find the latest production version. To do that, we will need to get the latest production Metadata and map that to a feature. It's doable though, so I can proceed with that.

As for
>If we follow the protocol of creating a new MV for each new feature and making them production ready at the same time then the answer to your question is yes.

Originally this did not need to be a production ready MV even if the feature is production, but I think we are now flipping this around and saying the feature if production ready iff the MV is production ready.

Reply:
> Originally this did not need to be a production ready MV even if the feature is production, but I think we are now flipping this around and saying the feature is production ready iff the MV is production ready.

If there is a 1-to-1 mapping from feature to MV, reasoning about production readiness in one place seems simpler?